### PR TITLE
test(integration): publisher self-service & ingest backend integration tests

### PR DESCRIPTION
**Pull request by bbernstein**
## Summary
- New `backend/src/__tests__/integration/` suite covers publisher lifecycle, self-service portal, admin controls, ingest edge cases, and auth/limits — all journey-shaped, all running against an in-memory DDB fake (no docker, no DynamoDB Local).
- Pins the PR #100 (sticky `published`) and PR #78 (self-disable retracts) regressions so they can't recur.
- 45 new tests, all green, ~14s wall-clock.

## CI
This PR does NOT modify `.github/workflows/build-and-test.yml`. The `push:` trigger that gates merges on this suite is added by PR #103 (`chore/ci-concurrency-dedupe`), which is already open and mergeable. Once #103 lands, every push to a branch will run the integration suite as part of `test-backend`.

## Manual follow-up after merge
Add `test-backend (24)`, `test-backend (25)`, `test-frontend (24)`, `test-frontend (25)` to required status checks under main's branch protection (Settings → Branches → main → Edit). This requires GitHub UI access; cannot be set from a workflow file.

## Test plan
- [x] \`npx jest integration/ --coverage=false\` (8 suites, 45 tests, all green)
- [x] \`npm run test:ci\` (full backend suite, 46 suites, 664 tests, all green)
- [ ] CI green on this PR
- [ ] CI red if PR #100's reconciler fix is reverted (regression catches it)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by bbernstein** [comment]
All hard CI checks green and `claude-review` complete. `mergeStateStatus: CLEAN`. No pending reviewers. Ready for your merge.

Note: this PR deliberately does NOT modify `.github/workflows/build-and-test.yml` because PR #103 already adds the `push:` trigger that gates the new integration suite. Land #103 first, or in either order — they are non-conflicting.

File: backend/src/__tests__/integration/harness/inMemoryDocClient.ts [hunk 1]
```diff
@@ -0,0 +1,854 @@
...
+        const bv = b[sortKey!] as string | number | undefined;
+        if (av === undefined || bv === undefined) return 0;
+        if (av < bv) return input.ScanIndexForward === false ? 1 : -1;
+        if (av > bv) return input.ScanIndexForward === false ? -1 : 1;
+        return 0;
+      });
+    }
+    return {
+      Items: candidates.map(i => JSON.parse(JSON.stringify(i))),
+      Count: candidates.length,
```
[the remark this answers is not in the document]
Comment (bbernstein) [reply backend/src/__tests__/integration/harness/inMemoryDocClient.ts]: Fixed in e03e1bb. doQuery and doScan now share an applyPagination helper that resumes from ExclusiveStartKey, applies Limit after filtering, and emits LastEvaluatedKey built from the table hash+sort keys only when more items remain. Added 4 tests in inMemoryDocClient.test.ts covering Limit < total, continuation, Limit ≥ total (no LEK), and hash-only Scan pagination.

File: backend/src/__tests__/integration/harness/inMemoryDocClient.ts [hunk 2]
```diff
@@ -0,0 +1,854 @@
...
+    const t = this.requireTable(input.TableName);
+    let hashKey = t.spec.hashKey;
+    let sortKey = t.spec.sortKey;
+    if (input.IndexName) {
+      const gsi = (t.spec.gsis ?? []).find(g => g.name === input.IndexName);
+      if (!gsi) throw new Error(`Unknown index '${input.IndexName}' on table '${t.spec.name}'`);
+      hashKey = gsi.hashKey;
+      sortKey = gsi.sortKey;
+    }
+    const all = Array.from(t.items.values());
```
[the remark this answers is not in the document]
Comment (bbernstein) [reply backend/src/__tests__/integration/harness/inMemoryDocClient.ts]: Fixed in e03e1bb. Removed the unused hashKey local and the misleading comment about hashKey ordering. DDB makes no ordering promise without a sort key, and no production code in this repo relies on hash-only-sorted results.

File: backend/src/__tests__/integration/harness/harness.ts [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+  };
+  const runIngestInProcess = async (payload: { singlePublisherId?: string } = {}) => {
+    // Refresh deps.now to reflect the current clock.
+    const deps: IngestDeps = { ...ingestDeps, now: now.now };
+    await runIngest(deps, { singlePublisherId: payload.singlePublisherId });
+  };
+  const invoker = new FakeLambdaInvoker(runIngestInProcess);
+
+  // Wire handler-level injection points. Each call captures the prior value
+  // so dispose() can restore.
```
[the remark this answers is not in the document]
Comment (bbernstein) [reply backend/src/__tests__/integration/harness/harness.ts]: Fixed in 0a02bfa. Updated the comment to accurately describe what dispose() does: setters overwrite module-level singletons, dispose() passes null which makes each handler lazily reconstruct fresh defaults on the next call (no save/restore).

File: backend/src/__tests__/integration/harness/actors.ts [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+const ADMIN_EMAIL = 'admin@test.chqcal.local';
+// Match harness.ts's pre-import env. Both agree on the same value so the
+// adminHandler-captured constant matches what we sign with.
+const ADMIN_JWT_SECRET = process.env.JWT_SECRET ?? 'test-admin-jwt-secret';
+
+import jwt from 'jsonwebtoken';
+
+function mintAdminJwt(email = ADMIN_EMAIL): string {
+  return jwt.sign({ email, name: 'Test Admin' }, ADMIN_JWT_SECRET, { expiresIn: '7d' });
+}
```
[the remark this answers is not in the document]
Comment (bbernstein) [reply backend/src/__tests__/integration/harness/actors.ts:163]: Fixed in 0a02bfa. Removed the divergent fallback. actors.ts now reads process.env.JWT_SECRET in an IIFE and throws if unset, so any divergence from adminHandler surfaces as a fast test failure rather than silent verification mismatch. integrationSetup.ts already sets JWT_SECRET before adminHandler.ts loads, so the captured value and the value we sign with are guaranteed to be identical.

File: backend/src/__tests__/integration/ingestEdgeCases.test.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    const evt = { id: 'e1', startDate: '2026-08-01T10:00:00Z', endDate: '2026-08-01T11:00:00Z', lastModified: '2026-06-01T00:00:00Z' };
+    h.feeds.set(publisherId, feedOk(publisherId, 'U', [evt]));
+    await h.actors.ingest.run();
+    const before = (await h.store.listForPublisher(publisherId))[0];
+
+    // Re-ingest same event with same lastModified → no diff applied to it.
+    await h.actors.ingest.run();
+    const after = (await h.store.listForPublisher(publisherId))[0];
+    expect(after.updatedAt).toBe(before.updatedAt); // unchanged → no write
+  });
```
[the remark this answers is not in the document]
Comment (bbernstein) [reply backend/src/__tests__/integration/ingestEdgeCases.test.ts]: Fixed in 44610af. The test asserts updatedAt is unchanged on re-ingest with the same lastModified — i.e. that the row is NOT rewritten. Renamed to "event row is not rewritten (updatedAt preserved)" to match the assertion.

File: backend/src/__tests__/integration/authAndLimits.test.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      'http://127.0.0.1/feed.json',
+      'http://169.254.169.254/feed.json',
+    ]) {
+      await expect(h.actors.publisher.apply({
+        name: 'Bad',
+        email: `bad-${Math.random()}@example.com`,
+        sourceUrl: url,
+        sourceType: 'json',
+      })).rejects.toMatchObject({ statusCode: 400 });
+    }
```
[the remark this answers is not in the document]
Comment (bbernstein) [reply backend/src/__tests__/integration/authAndLimits.test.ts:118]: Fixed in 44610af. Replaced Math.random() with the loop index i so the test produces deterministic email addresses (bad-0@, bad-1@, bad-2@) reproducibly across runs.

**Comment by bbernstein** [comment]
**All HIGH and MEDIUM items from claude-review round 2 have been addressed**, plus one LOW (G — GSI LEK doc note).

### HIGH
- **`deepEq` is now structural and key-order-independent** — replaced JSON.stringify-based comparison with a recursive deep-equality function that handles primitives, arrays, and objects with arbitrary key insertion order. Added a regression test that builds two objects with the same keys in different orders and asserts a ConditionExpression compares them as equal. (`ac28c16`)
- **Rate-limit retry uses a fresh email** — `burst11@example.com` was the email that just got rate-limited; the post-clock-advance retry now uses `burst12@example.com` so the test no longer depends on rate-check-before-DB-write ordering. Added a code comment explaining why. (`a934529`)

### MEDIUM
- **Token extraction goes through `tokenFromMagicLink()` helper** that throws a descriptive error if `token` is absent, replacing the seven `searchParams.get('token')!` non-null assertions across `lifecycle`, `selfService`, `adminControls`, `ingestEdgeCases`, and the actors smoke test. (`8580740`)
- **Removed dead `void selfDisable` import + misleading comment** in `harness.ts` — the import was never used; `portal._setSelfDisableDepsForTests` is what actually wires the handler. (`8580740`)

### LOW (done)
- **`applyPagination` GSI note** — added a comment near the LEK construction documenting that LEK currently only includes base-table keys and that's safe because no production query paginates over a GSI. (`d0bc6ad`)

### LOW (deferred)
- **E** (unused `currentHarness()` export) and **F** (consolidating duplicate `applyApprove*` helpers across three test files) are deferred — net-additive cleanups that don't affect correctness or readability of any individual file.

### Coverage gaps (deferred — net-additive scope expansion)
`cancelEmailChangeBySelf` journey coverage, expired-JWT rejection coverage, and broader DNS-guard cases (IPv6 loopback, RFC 1918) are net-additive scope expansions not in the original plan — they belong in a follow-up coverage PR.

### Test results
669 backend tests pass (up from 664 baseline — +5 new tests for the deepEq order-independence fix).

CI should re-run and go green. Reply once `mergeStateStatus: CLEAN` returns.

**Comment by bbernstein** [comment]
**Round 3 claude-review verifies all Round 2 fixes ✓.** `mergeStateStatus: CLEAN`. PR ready for merge.

### Total review-cycle summary

**Round 1 (6 Copilot inline comments):** all addressed in `e03e1bb`, `0a02bfa`, `44610af` — verified by claude-review round 2.

**Round 2 (claude-review surfaced 5 HIGH/MED items):** all addressed in `ac28c16`, `a934529`, `8580740`, `d0bc6ad` — verified by claude-review round 3.

**Round 3 (claude-review):** verified all round 2 fixes. Remaining items are net-additive polish, deferred to a follow-up PR:
- `h.mail.lastTo(email)!` non-null assertions (same pattern as the token fix; failure mode is a TypeError instead of a clean Error)
- Unused `currentHarness()` export in `harness.ts` (test-only API surface, no consumers)
- Undocumented `NOT` grammar gap in `inMemoryDocClient.ts` parser comment (no production code uses NOT)
- Coverage gaps for `cancelEmailChangeBySelf` journey, expired-JWT rejection, IPv6/RFC1918 DNS-guard cases (net-additive scope)

### Test results
669 backend tests pass (44 integration tests across 8 files, 17 in-memory DDB self-tests, 1 harness smoke, 1 actor smoke; +5 from round 2 deepEq order-independence test).

### Threads to mark as resolved (manual UI action)
- Round 1: 3198091739, 3198091770, 3198091783, 3198091788, 3198091799, 3198091805
- Round 2 / 3 findings were not inline threads (they were body-only review comments), so no extra UI resolution needed

All checks passing. `claude-review`: SUCCESS. `mergeStateStatus: CLEAN`. **PR ready for merge.**
